File: sdks/php/examples/TypedGreetNode.php
<?php

declare(strict_types=1);

namespace Blok\Blok\Examples;

use Blok\Blok\Node\TypedNode;
use Blok\Blok\Types\Context;

/** Typed input DTO for the typed-greet demo. */
final class TypedGreetInput
{
    public function __construct(public string $name, public int $repeat = 1)
    {
    }
}

/** Typed output DTO (declared so the output JSON Schema can be reflected). */
final class TypedGreetOutput
{
    public function __construct(public string $greeting, public int $length)
    {
    }
}

/** Typed greeting node demonstrating the SPEC-B TypedNode contract. */
final class TypedGreetNode extends TypedNode
{
    public function name(): string
    {
        return 'typed-greet';
    }

    public function description(): string
    {
        return 'Typed greeting (SPEC-B contract demo)';
    }

    protected function inputClass(): string
    {
        return TypedGreetInput::class;
    }

    protected function outputClass(): ?string
    {
        return TypedGreetOutput::class;
    }

    public function capabilityManifest(): ?array
    {
        return [
            'version' => '1',
            'classification' => 'agent-compatible',
            'effects' => [],
            'capabilities' => [],
            'secrets' => [],
            'determinism' => 'deterministic',
            'idempotency' => 'idempotent',
            'maturity' => 'stable',
            'resources' => [
                'maxDurationMs' => 5000,
                'maxInputBytes' => 4194304,
                'maxOutputBytes' => 4194304,
                'maxConcurrency' => 64,
            ],
        ];
    }

    protected function run(Context $ctx, object $input): mixed
    {
        /** @var TypedGreetInput $input */
        $repeat = $input->repeat > 0 ? $input->repeat : 1;
        $greeting = str_repeat('Hello, ' . $input->name, $repeat);

        return ['greeting' => $greeting, 'length' => strlen($greeting)];
    }
}

File: sdks/python3/examples/typed_greet_node.py
"""Typed greeting node demonstrating the SPEC-B @node contract."""

from __future__ import annotations

from pydantic import BaseModel

from blok import node
from blok.types.context import Context


class TypedGreetInput(BaseModel):
    name: str
    repeat: int = 1


class TypedGreetOutput(BaseModel):
    greeting: str
    length: int


@node(
    "typed-greet",
    "Typed greeting (SPEC-B contract demo)",
    capability_manifest={
        "version": "1",
        "classification": "agent-compatible",
        "effects": [],
        "capabilities": [],
        "secrets": [],
        "determinism": "deterministic",
        "idempotency": "idempotent",
        "maturity": "stable",
        "resources": {
            "maxDurationMs": 5000,
            "maxInputBytes": 4194304,
            "maxOutputBytes": 4194304,
            "maxConcurrency": 64,
        },
    },
)
def typed_greet(ctx: Context, input: TypedGreetInput) -> TypedGreetOutput:
    repeat = input.repeat if input.repeat > 0 else 1
    greeting = ("Hello, " + input.name) * repeat
    return TypedGreetOutput(greeting=greeting, length=len(greeting))
